typespec-extension/src/code-model-builder.ts

```diff
@@ -548,121 +515,196 @@ export class CodeModelBuilder {
+          }
+        }
       }
     }
-    if (apiVersionSameForAllClients) {
-      const serviceVersion = getServiceVersion(this.codeModel);
-      for (const client of this.codeModel.clients) {
-        client.serviceVersion = serviceVersion;
+  }
+
+  private buildSdkPathPathParameterForARM(): SdkPathParameter {
+    return {
+      kind: "path",
+      name: "endpoint",
+      isGeneratedName: true,
+      description: "Service host",
+      onClient: true,
+      urlEncode: false,
+      optional: false,
+      serializedName: "endpoint",
+      correspondingMethodParams: [],
+      type: {
+        kind: "string",
+        encode: "string",
+        decorators: [],
+        name: "string",
+        crossLanguageDefinitionId: "string",
+      },
+      isApiVersionParam: false,
+      decorators: [],
+      apiVersions: [],
+      crossLanguageDefinitionId: "endpoint",
+    };
+  }
+
+  private listSubClientsUnderClient(
+    client: SdkClientType<SdkHttpOperation>,
+    includeNestedOperationGroups: boolean,
+    isRootClient: boolean,
+  ): SdkClientType<SdkHttpOperation>[] {
+    const operationGroups: SdkClientType<SdkHttpOperation>[] = [];
+    for (const method of client.methods) {
+      if (method.kind === "clientaccessor") {
+        const subClient = method.response;
+        if (!isRootClient) {
+          // if it is not root client, append the parent client's name
+          subClient.name = this.removeClientSuffix(client.name) + this.removeClientSuffix(pascalCase(subClient.name));
+        }
+        operationGroups.push(subClient);
+        if (includeNestedOperationGroups) {
+          for (const operationGroup of this.listSubClientsUnderClient(subClient, includeNestedOperationGroups, false)) {
+            operationGroups.push(operationGroup);
+          }
+        }
       }
-    } else {
-      for (const client of this.codeModel.clients) {
-        const apiVersions = client.apiVersions;
-        if (apiVersions) {
-          client.serviceVersion = getServiceVersion(client);
+    }
+    return operationGroups;
+  }
+
+  private listServiceMethodsUnderClient(
+    client: SdkClientType<SdkHttpOperation>,
+    includeNestedServiceMethods: boolean,
+  ): SdkServiceMethod<SdkHttpOperation>[] {
+    const methods: SdkServiceMethod<SdkHttpOperation>[] = [];
+    for (const method of client.methods) {
+      if (includeNestedServiceMethods) {
+        if (method.kind === "clientaccessor") {
+          const client = method.response;
+          for (const method of this.listServiceMethodsUnderClient(client, includeNestedServiceMethods)) {
+            methods.push(method);
+          }
         }
       }
+      if (method.kind !== "clientaccessor") {
+        methods.push(method);
+      }
     }
+    return methods;
+  }
 
-    return clients;
+  private removeClientSuffix(clientName: string): string {
+    return clientName.endsWith("Client") ? clientName.slice(0, -6) : clientName;
   }
```

Comment:
Move these to util class.
It does not need to be a class method.

Also `-6` is magic number. Just do a `const clientSuffix = "Client";` then `slice(0, -clientSuffix.length)`.